cmd/kube-proxy/app/server.go

```diff
@@ -778,35 +778,42 @@ func (s *ProxyServer) birthCry() {
 	s.Recorder.Eventf(s.NodeRef, nil, api.EventTypeNormal, "Starting", "StartKubeProxy", "")
 }
 
-// detectNodeIP returns the nodeIP used by the proxier
+// detectNodeIPs returns the proxier's "node IP" or IPs, and the IP family to use if the
+// node turns out to be incapable of dual-stack. (Note that kube-proxy normally runs as
+// dual-stack if the backend is capable of supporting both IP families, regardless of
+// whether the node is *actually* configured as dual-stack or not.)
+
+// (Note that on Linux, the node IPs are used only to determine whether a given
+// LoadBalancerSourceRanges value matches the node or not. In particular, they are *not*
+// used for NodePort handling.)
+//
 // The order of precedence is:
-// 1. config.bindAddress if bindAddress is not 0.0.0.0 or ::
-// 2. the primary IP from the Node object, if set
-// 3. if no IP is found it defaults to 127.0.0.1 and IPv4
-func detectNodeIP(client clientset.Interface, hostname, bindAddress string) net.IP {
+//  1. if bindAddress is not 0.0.0.0 or ::, then it is used as the primary IP.
+//  2. if the Node object can be fetched, then its primary IP is used as the primary IP
+//     (and its secondary IP, if any, is just ignored).
+//  3. otherwise the primary node IP is 127.0.0.1.
+//
+// In all cases, the secondary IP is the zero IP of the other IP family.
+func detectNodeIPs(client clientset.Interface, hostname, bindAddress string) (v1.IPFamily, map[v1.IPFamily]net.IP) {
 	nodeIP := netutils.ParseIPSloppy(bindAddress)
 	if nodeIP.IsUnspecified() {
 		nodeIP = utilnode.GetNodeIP(client, hostname)
+		klog.InfoS("Detected node IP", "address", nodeIP.String())
```

Comment:
can it be nil here?

Reply:
yes. And actually, `utilnode.GetNodeIP` logs it itself anyway (on success) so we don't need to do anything